Fixes #36688 - Provide option to use wget for the new Register Host feature

goarsna opened the pull request:
The Register hosts page uses `curl` for registering hosts. This seems to be a good approach for EL based systems as most of these come with `curl` preinstalled. On Debian based systems this is not always the case (at least for recent versions of Debian and Ubuntu).

This feature request aims to provide an option to use `wget` for registering hosts as `wget` is as widely used as `curl`. By this system administrators that only have `wget` installed are not forced to also install curl anymore.

goarsna committed:
Fixes #36688 - Provide option to use wget for the new Register Host feature

goarsna commented:
I have opened a PR to document this in `foreman-documentation`:
https://github.com/theforeman/foreman-documentation/pull/2376

ekohl commented on app/controllers/api/v2/registration_commands_controller.rb:
```diff
@@ -21,6 +21,7 @@ class RegistrationCommandsController < V2::BaseController
         param :update_packages, :bool, desc: N_("Update all packages on the host")
         param :repo, String, desc: N_("Repository URL / details, for example for Debian OS family: 'deb http://deb.example.com/ buster 1.0', for Red Hat and SUSE OS family: 'http://yum.theforeman.org/client/latest/el8/x86_64/'")
         param :repo_gpg_key_url, String, desc: N_("URL of the GPG key for the repository")
+        param :use_wget, :bool, desc: N_("Use wget instead of curl")
```
Rather than a boolean parameter, would it make sense to use an enum that defaults to `curl` but also accepts `wget` just to be explicit?

goarsna replied:
I agree. I will change the boolean to an enum and replace the checkbox in the Host Registration page by a `FormSelect` field defaulting to `curl`.

goarsna replied:
Done :)

ekohl commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -16,7 +16,19 @@ def registration_args
+    useWget ? 'wget --no-verbose -O-' : 'curl -sS'
+  end
+
+  def useWget
```
snake_case is the convention.
```suggestion
  def use_wget
```

goarsna replied:
Thanks for the hint!

ekohl commented on app/controllers/api/v2/registration_commands_controller.rb:
```diff
@@ -21,6 +21,7 @@ class RegistrationCommandsController < V2::BaseController
         param :update_packages, :bool, desc: N_("Update all packages on the host")
         param :repo, String, desc: N_("Repository URL / details, for example for Debian OS family: 'deb http://deb.example.com/ buster 1.0', for Red Hat and SUSE OS family: 'http://yum.theforeman.org/client/latest/el8/x86_64/'")
         param :repo_gpg_key_url, String, desc: N_("URL of the GPG key for the repository")
+        param :download_utility, String, desc: N_("The download utility to use for host registration")
```
I think you can limit it with a validator. See https://github.com/Apipie/apipie-rails#parameter-description as well.

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -96,22 +106,36 @@ else
 <% end -%>
 
+<%
+if @download_utility == 'curl'
+  data_keyword = "--data"
+else
+  data_keyword = "--post-data"
+end
+
+if @download_utility == 'curl'
+  registration_command="curl --silent --show-error --request POST --cacert"
+else
+  registration_command="wget --no-verbose -O- --ca-certificate"
+end
```
Perhaps group the together?
```suggestion
if @download_utility == 'curl'
  data_keyword = "--data"
  registration_command="curl --silent --show-error --request POST --cacert"
else
  data_keyword = "--post-data"
  registration_command="wget --no-verbose -O- --ca-certificate"
end
```

Or another way:
```ruby
def build_command(utility, url, headers=[], data={})
  # ...
end
```

Then you can use it in both places.

goarsna commented:
Thanks @ekohl. I've incorporated your feedback and rebased my branch.

stejskalleos commented:
ok to test

stejskalleos commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -16,7 +16,27 @@ def registration_args
+  end
+
+  def download_utility
+    registration_params['download_utility']
```
What if the parameter is not set? By default the value should be a 'curl'

goarsna replied:
Added `curl` as default

stejskalleos commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -16,7 +16,27 @@ def registration_args
-    registration_params['insecure'] ? '--insecure' : ''
+    if registration_params['insecure']
+      if download_utility == 'curl'
+        ' --insecure'
```
Nit: Space at the beginning should be handled in the `command` method

goarsna replied:
Applied

stejskalleos commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -16,7 +16,27 @@ def registration_args
   end
 
   def insecure
```
Little refactoring with safe guard:

```ruby
return '' unless registration_params['insecure']&.presence

if download_utility == 'curl'
  '--insecure'
else
  '--no-check-certificate'
end
```

goarsna replied:
Thanks, this is a nice way to check the `insecure` flag. But regarding to some tests I did, the first line has to be `return '' unless registration_params['insecure']`. With this little change the results I get are

- `''` in case `insecure` is not set in the `registration_params`,
- `''` in case `insecure` is present in the `registration_params` an set to `false`, and
- in case  `insecure` is present in the `registration_params` an set to `true` the lines below are evaluated.

With the `&.presence` part I only get a `undefined method` error in case `insecure` is present in the `registration_params`.

stejskalleos commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -96,22 +106,32 @@ else
+  data_keyword = "--data"
+  registration_command="curl --silent --show-error --request POST --cacert"
+else
+  data_keyword = "--post-data"
```
Nit: `data_key` is sufficient in this case

goarsna replied:
Applied

stejskalleos commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/DownloadUtility.js:
```diff
@@ -0,0 +1,50 @@
+      id="reg_download_utility"
+      isDisabled={isLoading}
+    >
+      {_.map(DownloadUtilities, (key, value) => (
```
I'm not much JS proficient, what does the `_.map` do?

goarsna replied:
The `_` refers to the Lodash (and Underscore) JS libraries, which provide functions for manipulating arrays, lists, etc. I initially thought it is used in Foreman. But I saw it is only pulled in as dependency.

I now use the default `map()` function. I also switched to using a list for the `DownloadUtilities` as there is no need for using a dictionary.

goarsna commented:
Thanks @stejskalleos for your review, I finally made it to answer / adapt everything.

Please let me know if you would like to see any further changes.

sbernhard commented:
ok to test

stejskalleos commented:
@nofaralfasi can you take a look and do the review?

sbernhard commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -16,7 +16,25 @@ def registration_args
-    registration_params['insecure'] ? '--insecure' : ''
+    return '' unless registration_params['insecure']
+
+    if download_utility == 'curl'
```
What do you think to add something like this:
```
utility = {
 curl => { insecure => '--insecure', download => 'curl -sS },
wget => { insecure => '--no-check-certificate, download => 'wget --no-verbose -O -' }
}
```

This can be extended to other tools later on.

goarsna replied:
Thanks for that! I reworked the parts in `registration_commands.rb` and `global_registration.erb`

nofaralfasi commented:
When attempting to preview the 'Global Registration' template from the template edit page, I encounter the following error: `There was an error rendering the Global Registration template: undefined method '#[]' for NilClass::Jail (NilClass)`.

Additionally, could we merge the 16 commits into a single commit?

goarsna commented:
Hey @nofaralfasi, thanks for your feedback. I fixed the template, squashed the commits and rebased my branch.

goarsna commented:
Hey @nofaralfasi, again some changes.
in the first commit I rebased my branch to solve merge conflicts in the `global_registration.erb` introduced by [replacing apt-key for Ubuntu and Debian](https://github.com/theforeman/foreman/commit/6575dc039092bcb2bf5c9e2bc45139cb16f6e26e) .
Then I've added some tests to cover my changes.
And lastly @sbernhard and I discussed moving the dicts holding the `curl` and `wget` command patterns to one place. 

What I just noticed is that there is still a failing node test. I'll have a look at that.

Just ping me, in case I can / should squash the commits again. :)

goarsna commented:
Ok, something weird was happening to my local tests, but now the tests are fixed. :)

goarsna commented:
Hey there,
I rebased my branch due to merge conflicts and squashed the commits. Now the Jenkins unit tests fail. However the tests are working in my forklift. I tried both, calling `bundle exec rake jenkins:unit TESTOPTS="-v" --trace` and running the failing test file directly calling `bundle exec rake test TEST=test/controllers/unattended_controller_test.rb`.

What do you think @nofaralfasi?

nofaralfasi commented:
I don't know why the tests are failing. 
Also, the wget command doesn't work for me, whereas the curl command does.

```
[root@rhel8-8-testing-machine ~]# set -o pipefail && wget --no-verbose -O- --no-check-certificate 'http://192.168.121.1:3000/register?download_utility=wget&location_id=2&operatingsystem_id=13&organization_id=1&update_packages=false' --header 'Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJpYXQiOjE3MDU5MzQyOTksImp0aSI6IjRkOWJiYmQ4Yzg2OGUyODNlZWE3OGExZmFiMjZiNGZkZmEwNTRiZDAzMGZiOTU1NGNlNGQ2M2U3ZDc0YTEyZDciLCJleHAiOjE3MDU5NDg2OTksInNjb3BlIjoicmVnaXN0cmF0aW9uI2dsb2JhbCByZWdpc3RyYXRpb24jaG9zdCJ9.m_d_whF1xwve7KGDDDbrBRihBJMrAUpu9849ilAWd_M' | bash
2024-01-22 14:38:40 URL:http://192.168.121.1:3000/register?download_utility=wget&location_id=2&operatingsystem_id=13&organization_id=1&update_packages=false [3594] -> "-" [1]
#
# Running registration
#
http://192.168.121.1:3000/register:
2024-01-22 14:38:41 ERROR 422: Unprocessable Entity.
```

goarsna commented:
Hey @nofaralfasi,
thanks a lot for your reviews. I'll have a look at the error.

fyi: The docs for this PR got merged yesterday.

Lennonka commented:
The docs have been reverted. I've created a new PR to include the docs ^^

goarsna commented:
Hey there,
updated, rebased and did tests with Ubuntu 22.04 and Almalinux 8 that were successful. But I cannot tell atm what is causing the failing test. I can have a look at it again in calendar week 24, when I am back from vacation.

This PR also requires https://github.com/theforeman/foreman/pull/10179 to use wget in the host init config template.

goarsna commented:
After having a look at the failing tests again, I assume they are not related to my PR. :thinking: But I am a little bit clueless what's causing them. Any ideas, suggestions, thoughts or objections?

ofedoren commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -1,5 +1,6 @@
 module Foreman::Controller::RegistrationCommands
   extend ActiveSupport::Concern
+  include ProvisioningTemplatesHelper
```
Why is it being explicitly included here? I don't see any change that uses those helpers.

EDIT: I see now, it's for `download_utility` method...

ofedoren commented on app/services/foreman/renderer/scope/macros/helpers.rb:
```diff
@@ -4,6 +4,7 @@ module Scope
       module Macros
         module Helpers
           include Foreman::Renderer::Errors
+          include ProvisioningTemplatesHelper
```
Almost the same question as before. Moreover, we don't include view helpers in macros.

goarsna replied:
May I ask how view helpers are defined? Or better to say, how can I determine if a module / class is a view helper?

ofedoren replied:
> how can I determine if a module / class is a view helper

They land in app/helpers and required/included automatically to be used within erb views.

goarsna replied:
Thanks a lot!

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/DownloadUtility.js:
```diff
@@ -0,0 +1,47 @@
+
+import { translate as __ } from '../../../../../common/I18n';
+
+export const DownloadUtilities = ['curl', 'wget'];
```
Add `wget` to https://github.com/theforeman/foreman/blob/develop/.eslintrc#L17, so GA doesn't spam with warnings.

ofedoren commented on app/helpers/provisioning_templates_helper.rb:
```diff
@@ -72,4 +72,25 @@ def how_templates_are_determined
      (link_to _("Operating System"), operatingsystems_path))).html_safe)
   end
+
+  def download_utility(utility)
```
I'd vote for extraction of this method. It's not tied to neither views nor erb macros. If there is no better place, I'd say we can bare with a custom module with this method or maybe include it in general helpers such as https://github.com/theforeman/foreman/blob/develop/lib/foreman.rb

Since it's a frozen hash, it could be a constant instead of a method.

goarsna replied:
Moved to `lib/foreman.rb` as constant

ofedoren commented on app/services/foreman/renderer/scope/macros/helpers.rb:
```diff
@@ -155,6 +156,32 @@ def truthy?(value = nil)
+          end
+
+          apipie :method, 'Formats post parameters for usage with either curl or wget.' do
+            required :params, [String], desc: 'The post parameters to format.'
```
IIRC:
```suggestion
            required :params, Array, of: [String], desc: 'The post parameters to format.'
```

goarsna commented:
Thanks @ofedoren for your review! Again the failing Katello tests...

goarsna commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -132,22 +135,23 @@ echo "#"
+  params << "packages=#{shell_escape(@packages)}" if @packages.present?
+  params << "update_packages=#{@update_packages}" unless @update_packages.nil?
+-%>
+UUID=$(subscription-manager identity | grep --max-count 1 --only-matching '\([[:xdigit:]]\{8\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{12\}\)')
```
Note to myself: Spaces missing
```suggestion
  UUID=$(subscription-manager identity | grep --max-count 1 --only-matching '\([[:xdigit:]]\{8\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{4\}-[[:xdigit:]]\{12\}\)')
```

goarsna commented:
Hey @sbernhard, as discussed in our call last week I moved the construction of the "curl" and "wget" commands completely to helper methods.
I will also have a look at unifying the two methods in one generic method for creating web requests in templates. But I have to see what the result looks like, considering all the cases and the formatting.

ofedoren commented on lib/foreman.rb:
```diff
@@ -45,4 +45,24 @@ def self.input_types_registry
     SettingRegistry.instance
   end
+
+  DOWNLOAD_UTILITIES =
```
@goarsna, I'm so sorry for suggesting you to make it a constant even though it makes sense... We're going to switch into Zeitwerk, which ideally wants a separate file for each constant, so let's not create a nested constant, which will be used globally within this file (even though it will still work). Just to make things simpler, let's rather convert it into a method whenever you're going to make another change within the PR.

goarsna replied:
No problem :) But would `lib/foreman.rb` still be the correct place to store the `download_utilities` method?

ofedoren replied:
Unless someone else objects, yeap.

goarsna commented on app/services/foreman/renderer/scope/macros/helpers.rb:
```diff
@@ -155,6 +155,33 @@ def truthy?(value = nil)
+            optional :indent, String, desc: 'Number of spaces the returned command should be indented.'
+            optional :skip1, [true, false], desc: 'True if the first line should not be indented.'
+            returns String, desc: 'The web request.'
+            example '...'
```
Note for myself: Example missing

goarsna commented:
Hey @ofedoren and @sbernhard, I (hopefully) addressed everything that was requested and fixed the tests again (despite those often failing katello tests... :sweat_smile:). And I again did test some tests: I registered an Ubuntu host using curl and wget and the command in the register host page worked as well as the rendered global_registration template with both, curl and wget.

Is there anything left from your side or are we good to merge this? :)

goarsna commented:
I received a good suggestion by @sbernhard: Add tests for the `generate_web_request method` :grin:  => Done

goarsna commented:
I did some other end-to-end tests with `bash -x` and also tested the host_init_config template with the changes from https://github.com/theforeman/foreman/pull/10179.
Thereby I noticed that https://github.com/theforeman/foreman/pull/9983 added installation of curl to the global_registration template. => Removed

goarsna commented:
In further downstream tests @dosas discovered that `wget` by default enables HSTS support whereas curl does not. This can lead to errors (for example) in case a repository is added to use during host registration which is secured with a self signed certificate.

To stay consistent accross `curl` and `wget` we added the `--no-hsts` flag to the `wget` commands. The `--no-hsts` flag has no effect on https connections.

sbernhard commented:
The failed tests are not related.

Any objections to get this change in @stejskalleos / @ekohl / @ofedoren ?

sbernhard reviewed (dismissed):
Thanks @goarsna

ofedoren commented on app/services/foreman/renderer/scope/macros/helpers.rb:
```diff
@@ -155,6 +155,32 @@ def truthy?(value = nil)
           end
+
+          apipie :method, 'Generate a web request' do
+            required :utility, String, desc: 'The utility to use for the web request ("curl" or "wget").'
+            required :url, String, desc: 'The URL for the web request.'
+            optional :ssl_ca_cert, String, desc: 'The path to the SSL certificate.'
+            optional :headers, Array, of: String, desc: 'Array containing the headers for the web request starting with "--header".'
+            optional :params, Array, of: String, desc: 'Array containing the POST parameters for the web request.'
+            optional :output_file, String, desc: 'The path were the result of the web request will be stored.'
```
There will still be a minor issue (bug in ApipieDSL), but it will still be more correct:
```suggestion
            keyword :utility, String, desc: 'The utility to use for the web request ("curl" or "wget").'
            keyword :url, String, desc: 'The URL for the web request.'
            keyword :ssl_ca_cert, String, desc: 'The path to the SSL certificate.', default: nil
            keyword :headers, Array, of: String, desc: 'Array containing the headers for the web request starting with "--header".', default: nil
            keyword :params, Array, of: String, desc: 'Array containing the POST parameters for the web request.', default: nil
            keyword :output_file, String, desc: 'The path were the result of the web request will be stored.', default: nil
```

goarsna replied:
What minor issue do you mean?
I found out that with this change the documentation of the `generate_web_request` method now states that all parameters are optional and have the default value `nil` whereas with `required` and `optional` the parameters appeared correctly in the documentation. Is this the bug you mean?

ofedoren replied:
Yeap. The suggested changes are still valid, it's just I need to update ApipieDSL with a fix for that misleading :/

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/General.js:
```diff
@@ -111,6 +122,7 @@ General.defaultProps = {
   operatingSystemId: undefined,
   operatingSystemTemplate: undefined,
   smartProxyId: undefined,
+  downloadUtility: undefined,
```
If it's semi required and actually expects one of the supported utils, maybe it should be `curl` by default instead?

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/General.js:
```diff
@@ -97,6 +106,8 @@ General.propTypes = {
   handleInsecure: PropTypes.func.isRequired,
   handleInvalidField: PropTypes.func.isRequired,
   isLoading: PropTypes.bool.isRequired,
+  downloadUtility: PropTypes.oneOf(['curl', 'wget']),
```
I don't think there will any more supported utils, but we can reuse `DownloadUtilities` constant here, so if one is updated, this is autoupdated as well.

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/DownloadUtility.js:
```diff
@@ -0,0 +1,47 @@
+
+DownloadUtility.propTypes = {
+  downloadUtility: PropTypes.oneOf([
+    DownloadUtilities.curl,
+    DownloadUtilities.wget,
```
Does this actually work? Shouldn't this be `PropTypes.oneOf(DownloadUtilities)`?

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/DownloadUtility.js:
```diff
@@ -0,0 +1,47 @@
+};
+
+DownloadUtility.defaultProps = {
+  downloadUtility: DownloadUtilities.curl,
```
```suggestion
  downloadUtility: DownloadUtilities[0],
```

ofedoren commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/index.js:
```diff
@@ -88,6 +89,9 @@ const RegistrationCommandsPage = () => {
   const [repoDataInternal, setRepoDataInternal] = useState([]);
   const [invalidFields, setInvalidFields] = useState([]);
+  const [downloadUtility, setDownloadUtility] = useState(
+    DownloadUtilities.curl
```
ditto

ofedoren reviewed (commented):
Mostly minor issues:

ofedoren commented:
Some js tests failed, could you please run `npm run test -- -u` to regenerate the snapshots?

goarsna commented:
Hey @ofedoren, thanks again for your review.
I left [one comment](https://github.com/theforeman/foreman/pull/9808#discussion_r1695055472) above and added missing Apipie documentation for the `download_utility` in the [registration_controller.rb](https://github.com/theforeman/foreman/compare/7c56ea6ba25f3b7b00ce584ca902d2c27b228e86..5ad1ce7f92f9c8f51f40f918e716994a99d7682e) :see_no_evil:

sbernhard commented on app/services/foreman/renderer/scope/macros/helpers.rb:
```diff
@@ -155,6 +155,32 @@ def truthy?(value = nil)
+              command << utility[:output_pipe]
+            end
+            headers&.each { |header| command << header }
+            utility[:format_params].call(params).each { |param| command << param } if params
```
We need to fix this because it generated (in case of curl):
`--data "packages=Paket1\ Paket2" which would mean it tries to install a package with the name "Paket1 Paket2" (which does not work because 2 packages are meant")

goarsna replied:
Thanks Bernhard, I removed the quotes around the post parameters. This made it necessary to escape the ampersands in the `wget` post parameter.
In my tests installing multiple packages during host registration now worked as expected.

goarsna replied:
@ofedoren Any thoughts about the solution with removing the quotes?

goarsna replied:
Maybe I should provide some additional information :sweat_smile: 

Post parameters have to be provided differently for `curl` (one `--data` argument for each post parameter) and `wget` (one `--post-data` argument containing all post parameters). Initially I removed the quotes of the post parameters in the Global Registration template to unify them and when formatting the post parameters for the `curl` or `wget` command respectively I added quotes around each post parameter in the case of `curl` and around the whole `--post-data` argument in the case of `wget`. But by this I added quotes also to those post parameters that where not quoted before (like the `packages` post parameter) which is causing the described problem.
While investigating the problem with installing multiple packages, I was thinking about whether it even makes sense to quote any of the post parameters as the parameters that could cause problems are already escaped. So my new approach is to not quote the parameters and escape them correctly.

sbernhard reviewed (changes requested):
--data "packages=Paket1\ Paket2" <- this need to be fixed. see comment above

goarsna commented:
I talked with @sbernhard and we were wondering why there are no snapshot tests for the Global Registration template. I am currently working on this and added an initial approach.

goarsna commented:
Rebased

goarsna commented:
Ok, I tried to add sensible and reasonable snapshot tests for the Global Registration template with various parameters set. But this consumes more time than expected.
As I do not want to block this PR with it, I'll move adding of snapshot tests for the Global Registration template to a separate PR.

How do we want to proceed with this PR? I hope it can be merged soon. :)

sbernhard reviewed (approved):
Thanks!

sbernhard commented:
@ekohl  / @stejskalleos you requested changes. Can you please check again. Thanks.

maximiliankolb commented:
@stejskalleos Friendly reminder to please re-review this PR.
